include/mesos/mesos.proto

```diff
@@ -3328,6 +3328,9 @@ message TTYInfo {
  */
 message ContainerInfo {
   // All container implementation types.
+  // For each type there should be a field in the ContainerInfo itself
+  // with exactly matching name in the lowercase.
```

Comment:
I just realized that the protobuf docs removed mention of their union trick a few months ago 😱 
https://github.com/protocolbuffers/protobuf/issues/452

Here is a snapshot of the union trick documentation:
https://web.archive.org/web/20170715111519/https://developers.google.com/protocol-buffers/docs/techniques

Reply:
Alternatively, we can look at the replacement:
https://developers.google.com/protocol-buffers/docs/proto#oneof

Reply:
We looked at `oneof` while writing the patches, but we cannot use it here because it breaks backwards compatibility.

Reply:
Unfortunately, there are two problems with converting this external interface to `oneof`.
- First, it would make the code that sets both `docker` and `mesos` silently do strange things (instead of failing). I cannot prove that there is no such code outside of Mesos.
- Second, we will still need `ContainerInfo::Type` and thus the validation. There are a lot of tests that only do `set_type(Type::MESOS)` but set neither `docker` nor `mesos`. According to the comment above the `ContainerInfo` definition, this is a perfectly valid way to set up a `MESOS` container without a filesystem isolation. I would expect such code to also exist in production outside of Mesos itself.

It would have been a right decision to use `oneof` when this interface was initially introduced (and do without the `type` field), but now it is too late:(